### ISSUE TITLE
store.sagaTask incompatible with typescript?

### ISSUE THREAD

**Issue by tunesmith** [bmealhouse/next-redux-saga#64]
Hi, after following these instructions:

```
- sagaMiddleware.run(rootSaga)
+ store.sagaTask = sagaMiddleware.run(rootSaga)
```

and renaming my store.js to store.ts, I found that the first line wouldn't work because of an unresolved promise (see #55), and the second line wouldn't work because Typescript doesn't like dynamically creating the "sagaTask" field on the store object.

Is there another way to ensure the sagaMiddleware completes without assigning it to store.sagaTask?

**Comment by bbortt** [comment]
As long as you do not want to write all the boilerplate code yourself, not there is not. It comes from this line in `next-redux-saga/index.js`
```js
import {END} from 'redux-saga'
[...]
store.dispatch(END)
await store.sagaTask.toPromise()
```
and makes sure you wait until all saga tasks have completed, before rendering the page.
Two points to consider:
* SSR is rather outdated, and I personally don't use `next-redux-saga` any more. `next.js` has migrated over to static page (pre-) rendering and you would therefore just use a classical `react`-`redux`-`saga` setup.
* Besides, just "renaming" your files to `.ts` instead of `.js` wont help much. You're in a JavaScript or TypeScript project now? When using `react` I personally recommend taking a look at `flow` type checker. ;)

**Comment by tunesmith** [comment]
@bbortt Would you be willing to explain what you mean by SSR being outdated? As more of a backend guy I'm still learning specifics here. I understand the basic differences between CSR, SSR, and SSG, but by static page rendering do you mean SSG? How does saga relate to SSG? For my purposes, I definitely want client-side interactivity (that uses api data sources) after the content is loaded. Since I'm using redux and saga, that made me think I needed next-redux-saga.

**Comment by bbortt** [comment]
Hi @tunesmith 

Sure thing, I will try!

By static page rendering I mean you run `next build` and `next export` and `next.js` will render your basic pages into different HTML files. This has several advantages over server side rendering (that is `next start`), for example you won't need a server.

From [this Auth0 article](https://auth0.com/blog/ultimate-guide-nextjs-authentication-auth0/):
> This model has several advantages when it comes to hosting. Static hosting sites (like Now, Amazon S3, Azure Blob Storage, Netlify, and others) are battle-tested, inexpensive, but more importantly, they are extremely fast and play well with CDNs.

The goal of this project (`next-redux-saga`) is to await side effects when server side rendering in order to send the client a fully filled (html) page. Hovewer `js` or rather `react` will catch up clientside anyways. So there's no real advantage (in most cases) over static pre-rendered files. Clients are pretty fast these days and there's not much of a wait time until `react` caught up client side.

Besides that, #55 was resolved and closed. If you still experience problems, I would recommend setting up a little example project and push it to GitHub. I could take a look at it then ;)

**Comment by tunesmith** [comment]
Ok, thanks! I see what you mean now - that more seems a case for client-side rendering (it was confusing me that NextJS was calling this SSG) over server-side rendering. I agree with you about most cases, but in the cases where SSR is still warranted, it seems that is where next-redux-saga is useful (when using nextjs, SSR, and saga).

So this issue is specifically about store.sagatask. I understand that using store.sagatask is the fix for #55 . But since NextJS has first class support for typescript now, it looked this approach might be a bug since store.sagatask (assigning sagatask to store) causes typescript errors. 

In other words, the [recommended approach](https://github.com/zeit/next.js/blob/canary/examples/with-redux-saga/store.js) of 

```
store.sagaTask = sagaMiddleware.run(rootSaga)
```

causes typescript errors when attempted in a typescript file. Since NextJS has built in support for typescript, it would be nice if next-redux-saga worked with typescript as well. Not sure if this is a bug report or a feature request. :)

**Comment by bjoluc** [comment]
@tunesmith You can trick TypeScript to accept the `sagaTask` assignment by writing
```typescript
(store as any).sagaTask = sagaMiddleware.run(rootSaga)
```
Maybe, an alternative would be to augment the redux store type, but since you won't need to manually access `sagaTask` this wouldn't be more helpful than tricking TypeSript at this place.

**Comment by bjoluc** [comment]
Two options here:

* a) `next-redux-saga` provides a separate TypeScript example with the `(store as any)` cheat
* b) `next-redux-saga` augments the redux store type accordingly

@bbortt @bmealhouse What do you think?

**Comment by bbortt** [comment]
No problem @tunesmith. Sure you can cast it as `any` but this is rather unsafe. I think the `redux` store is typed in TypeScript. A better way would be to write & assign your own type. An extension of the store type.

@bjoluc I honestly dont know how TypeScript types will be exported when packing the project. But it's not a too big task to rewrite the project in `.ts`.

**Comment by bjoluc** [comment]
> Sure you can cast it as any but this is rather unsafe.

You won't get an error if you misspell `.sagaTask = sagaMiddleware.run(rootSaga)`, but that should be it, shouldn't it?

> I think the redux store is typed in TypeScript. A better way would be to write & assign your own type. An extension of the store type.

Yes, `redux` provides its own type definitions. With `strict` type checking, however, providing an extended `store` where a plain `store` is expected will result in errors, AFAIK. Hence the augmentation thing...

Oh right, there's nothing about TypeScript in this repo, whoops! You're right, should be rather simple. If this is intended, I may be able to do a PR within the next month.

**Comment by bbortt** [comment]
Sure, I'd be happy to accept pull-requests at any time :)
